site/src/components/pages/docs/search/search-list-category.tsx

```diff
@@ -0,0 +1,82 @@
+import { Box, Grid, Typography, useTheme } from "@mui/material";
+import React, { useMemo, useState } from "react";
+
+import { AlgoliaHighlightResult, AlgoliaResult } from "./index";
+import SearchHeading from "./search-list-heading";
+import SearchItem from "./search-list-item";
+
+interface SearchItemCategoryProps {
+  title: string;
+  icon: JSX.Element;
+  searchResults: AlgoliaResult[];
+  getHighlight: (highlight: AlgoliaHighlightResult) => string;
+  closeModal?: () => void;
+}
+
+const SearchListCategory: React.VoidFunctionComponent<
+  SearchItemCategoryProps
+> = ({ title, icon, searchResults, getHighlight, closeModal }) => {
+  const theme = useTheme();
+  const [sections, setSections] = useState<AlgoliaResult[]>([]);
+  const [items, setItems] = useState<AlgoliaResult[]>([]);
+
+  useMemo(() => {
+    setSections(
+      searchResults.filter(
+        (result) => result._highlightResult?.title?.matchLevel !== "none",
+      ),
+    );
+    setItems(
+      searchResults.filter(
+        (result) => result._highlightResult?.title?.matchLevel === "none",
+      ),
+    );
+  }, [searchResults]);
```

Comment:
To expand on what @nathggns said on this – `sections` and `items` do not need to be in state at all.

They are derived from the prop `searchResults`, which is going to cause a re-render whenever it changes. With the existing approach, when `searchResults` updates the component is going to re-render, and _then_ we call `setSections/Items` with the new values, and then the component renders _again_ because state values have changed. There's no need for this – we only want a single re-render when the search results changes. Whenever we have a value that is always going to change at the same time as an existing prop or state value and is simply a result of applying some formula to that existing value, we don't need to store it in its own state variable.

We could just calculate these values directly on every render: 

```suggestion
  const sections = searchResults.filter(
    (result) => result._highlightResult?.title?.matchLevel !== "none",
  );
  const items = searchResults.filter(
    (result) => result._highlightResult?.title?.matchLevel === "none",
  );
```

with this approach they'll be calculated on every render, even if `searchResult` hasn't changed. e.g. as soon as someone types a letter in the search box, that updates state in `ModalSearch`, and this is a child of `ModalSearch` so it re-renders and these are run again, despite `searchResults` not having changed. 

We can use `useMemo` to avoid this, and only re-calculate sections/items when `searchResults` changes, as below. There probably isn't any practical difference in this case since the search results list is very short and filtering it is not expensive.

```suggestion
  const [sections, items] = useMemo(
    () => [
      searchResults.filter(
        (result) => result._highlightResult?.title?.matchLevel !== "none",
      ),
      searchResults.filter(
        (result) => result._highlightResult?.title?.matchLevel === "none",
      ),
    ],
    [searchResults],
  );
```

Reply:
Thank you for the great insight, @CiaranMn!